packages/eslint-plugin-next/lib/rules/no-page-custom-font.js

```diff
@@ -22,17 +38,91 @@ module.exports = {
+          node.declaration.superClass.name === documentImportName
+        ) {
+          localDefaultExportName = node.declaration.id.name
+          exportDeclarationType = 'ClassDeclaration'
+        }
+      },
+
       JSXOpeningElement(node) {
-        const documentClass = context
-          .getAncestors()
-          .find(
-            (ancestorNode) =>
-              ancestorNode.type === 'ClassDeclaration' &&
-              ancestorNode.superClass &&
-              ancestorNode.superClass.name === documentImportName
+        if (node.name.name !== 'link') {
+          return
+        }
+
+        const ancestors = context.getAncestors()
+
+        // if `export default <name>` is further down within the file after the
+        // currently traversed component, then `localDefaultExportName` will
+        // still be undefined
+        if (!localDefaultExportName) {
+          // find the top level of the module
+          const program = ancestors.find(
+            (ancestor) => ancestor.type === 'Program'
           )
 
-        if (documentClass || node.name.name !== 'link') {
+          // go over each token to find the combination of `export default <name>`
+          for (let i = 0; i <= program.tokens.length - 1; i++) {
+            if (localDefaultExportName) {
+              break
+            }
+
+            const token = program.tokens[i]
+
+            if (token.type === 'Keyword' && token.value === 'export') {
+              const nextToken = program.tokens[i + 1]
+
+              if (
+                nextToken &&
+                nextToken.type === 'Keyword' &&
+                nextToken.value === 'default'
+              ) {
+                const maybeIdentifier = program.tokens[i + 2]
+
+                if (maybeIdentifier && maybeIdentifier.type === 'Identifier') {
+                  localDefaultExportName = maybeIdentifier.value
+                }
+              }
+            }
+          }
+        }
+
+        const parentComponent = ancestors.find((ancestor) => {
+          // export default class ... extends ...
+          if (exportDeclarationType === 'ClassDeclaration') {
+            return (
+              ancestor.type === exportDeclarationType &&
+              ancestor.superClass &&
+              ancestor.superClass.name === documentImportName
+            )
+          }
+
+          // export default function ...
+          if (exportDeclarationType === 'FunctionDeclaration') {
+            return (
+              ancestor.type === exportDeclarationType &&
+              ancestor.id.name === localDefaultExportName
+            )
+          }
+
+          // function Name() {} export default XY
+          // class Name extends Document; export default Name
+          return ancestor.id && ancestor.id.name === localDefaultExportName
+        })
+
+        // file starts with _document and this <link /> is within the default export
+        if (is_Document && parentComponent) {
```

Comment:
Forgive me if I'm missing something here, but can we not just check if the file starts with `_document` in the beginning of `JSXOpeningElement` and return immediately if it is (since this rule is about flagging custom fonts in specific pages)?

```js
if (is_Document) {
  return
}
```

Reply:
I'm not 100% sure how the webpack plugin works specifically, here's the scenario I had in mind: if you group your `<link />`s into a separate component defined within `_document`, let's call it `<Links />` and use that within `Document`, will it work just the same? Because if not, then that component should be warned on.

Reply:
Do you mean something like this?

```jsx
function Links() {
  return (
    <>
      <link ... />
      <link ... />
    </>
  )
}

export default function Document() {
  return (
    <Html>
      <Head>
         <Links /> // links subcomponent
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  )
}
```

Have you tested this scenario by any chance? Because I think the behavior would work the same 🤔

Reply:
You're right, I should've just checked myself. The behaviour is indeed different - given

```js
import { Head, Html, Main, NextScript } from "next/document";

export default function Document() {
  return (
    <Html>
      <Head>
        <Links />
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}

function Links() {
  return (
    <>
      <link
        href="https://fonts.googleapis.com/css2?family=Krona+One&display=swap"
        rel="stylesheet"
      />
      <link
        href="https://fonts.googleapis.com/css2?family=Open+Sans&display=swap"
        rel="stylesheet"
      />
    </>
  );
}

```
the result will **not** be optimized.

Reply:
I've adjusted the warning for this case, exteded tests as well as the docs. Let me know what you think!

Reply:
Thanks for testing! This looks great, and we may need to look into ensuring font optimization still happens even if a separate component is used. But this works great in the meantime :)